File: clients/rust/src/live_chat.rs
mod proto;

use proto::chat_package::Message;
use proto::chat_package::chat_service_client::ChatServiceClient;
use std::error::Error;
use std::time::Duration;
use tokio::time::sleep;
use tokio_stream::{StreamExt, wrappers::ReceiverStream};
use tonic::transport::{Channel, ClientTlsConfig};

const SERVER_ADDR: &str = "https://grpc-sample-35975833932.southamerica-west1.run.app"; // Demo server

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let pid = std::process::id();
    // Connect to the gRPC server
    rustls::crypto::ring::default_provider().install_default().expect("Failed to install rustls crypto provider");
    let tls = ClientTlsConfig::new()
        .with_native_roots()
        .assume_http2(true);    
    let channel = Channel::from_static(SERVER_ADDR).tls_config(tls)?.connect().await?;
    let mut grp_service = ChatServiceClient::new(channel);

    let (tx, rx) = tokio::sync::mpsc::channel(128);

    tokio::spawn(async move {
        for i in 0..100 {
            sleep(Duration::from_secs(1)).await;
            let message = Message {
                content: format!("Hi from PID {:?}. This is test #{:?}", pid, i).to_string(),
            };
            println!("Streaming: {:?}", message);
            tx.send(message).await.unwrap();
        }
    });

    let mut in_stream = grp_service
        .live_chat(ReceiverStream::new(rx))
        .await?
        .into_inner();

    while let Some(result) = in_stream.next().await {
        println!("Got from stream: {:?}", result);
    }
    println!("Stream ended");

    Ok(())
}

File: clients/go/live_chat.go
package main

import (
	"context"
	"crypto/tls"
	"fmt"
	"io"
	"log"
	"os"
	"sync"
	"time"

	proto "grpc-sample-client/protos"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
)

func sending(stream proto.ChatService_LiveChatClient, wg *sync.WaitGroup) {
	defer wg.Done()
	var pid = os.Getpid()
	for i := range 100 {
		var request = proto.Message{Content: fmt.Sprintf("Hi from PID %d. This is test %d", pid, i)}
		log.Printf("Streaming: %v", &request)
		err := stream.Send(&request)
		if err != nil {
			log.Fatalf("Error streaming: %v", err)
		}
		time.Sleep(2 * time.Second)
	}
}

func receiving(stream proto.ChatService_LiveChatClient, wg *sync.WaitGroup) {
	defer wg.Done()
	for {
		resp, err := stream.Recv()
		if err == io.EOF {
			break // End of stream
		}
		if err != nil {
			log.Fatalf("Error receiving stream: %v", err)
		}
		log.Printf("Response: %v", resp)
	}
}

func main() {
	var addr = "grpc-sample-35975833932.southamerica-west1.run.app" // Demo server
	var tlsConfig = &tls.Config{InsecureSkipVerify: true}
	// Set up a connection to the server.
	conn, err := grpc.NewClient(addr, grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)))
	if err != nil {
		log.Fatalf("did not connect: %v", err)
	}
	defer conn.Close()
	gRPCService := proto.NewChatServiceClient(conn)

	// Contact the server and print out its response.
	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
	defer cancel()

	stream, err := gRPCService.LiveChat(ctx)
	if err != nil {
		log.Fatalf("could not greet: %v", err)
	}

	var wg sync.WaitGroup

	wg.Add(2)
	go sending(stream, &wg)
	go receiving(stream, &wg)

	wg.Wait()
	log.Println("Bye")
}
